README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# torchopt

<!-- badges: start -->

[![R-CMD-check](https://github.com/e-sensing/torchopt/workflows/R-CMD-check/badge.svg)](https://github.com/e-sensing/torchopt/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/torchopt)](https://cran.r-project.org/package=torchopt)
[![Software Life
Cycle](https://img.shields.io/badge/lifecycle-experimental-yellow.svg)](https://lifecycle.r-lib.org/articles/stages.html)
[![Software
License](https://img.shields.io/badge/license-Apache%202-2--green)](https://www.apache.org/licenses/LICENSE-2.0)

<!-- badges: end -->

The `torchopt` package provides R implementation of deep learning
optimizers proposed in the literature. It is intended to support the use
of the torch package in R.

## Installation

Installing the CRAN (stable) version of `torchopt`:

``` r
install.packages("torchopt")
```

Installing the development version of `torchopt` do as :

``` r
library(devtools)
install_github("e-sensing/torchopt")
```

    #> Warning: package 'torch' was built under R version 4.1.3

## Provided optimizers

`torchopt` package provides the following R implementations of torch
optimizers:

-   `optim_adamw()`: AdamW optimizer proposed by Loshchilov & Hutter
    (2019). Converted from the `pytorch` code developed by Collin
    Donahue-Oponski available at
    <https://gist.github.com/colllin/0b146b154c4351f9a40f741a28bff1e3>

-   `optim_adabelief()`: Adabelief optimizer proposed by Zhuang et al
    (2020). Converted from the authors’ PyTorch code:
    <https://github.com/juntang-zhuang/Adabelief-Optimizer>.

-   `optim_adabound()`: Adabound optimizer proposed by Luo et al.(2019).
    Converted from the authors’ PyTorch code:
    <https://github.com/Luolc/AdaBound>.

-   `optim_adahessian()`: Adahessian optimizer proposed by Yao et
    al.(2021). Converted from the authors’ PyTorch code:
    <https://github.com/amirgholami>.

-   `optim_madgrad()`: Momentumized, Adaptive, Dual Averaged Gradient
    Method for Stochastic Optimization (MADGRAD) optimizer proposed by
    Defazio & Jelassi (2021). The function is imported from
    [madgrad](https://CRAN.R-project.org/package=madgrad) package and
    the source code is available at <https://github.com/mlverse/madgrad>

-   `optim_nadam()`: Incorporation of Nesterov Momentum into Adam
    proposed by Dozat (2016). Converted from the PyTorch site
    <https://github.com/pytorch/pytorch>.

-   `optim_qhadam()`: Quasi-hyperbolic version of Adam proposed by Ma
    and Yarats(2019). Converted from the code developed by Meta AI:
    <https://github.com/facebookresearch/qhoptim>.

-   `optim_radam()`: Rectified verison of Adam proposed by Liu et al.
    (2019). Converted from the PyTorch code
    <https://github.com/pytorch/pytorch>.

-   `optim_swats()`: Optimizer that switches from Adam to SGD proposed
    by Keskar and Socher(2018). Converted from the `pytorch` code
    developed by Patrik Purgai: <https://github.com/Mrpatekful/swats>

-   `optim_yogi()`: Yogi optimizer proposed by Zaheer et al.(2019).
    Converted from the `pytorch` code developed by Nikolay Novik:
    <https://github.com/jettify/pytorch-optimizer>

## Optimization test functions

You can also test optimizers using optimization [test
functions](https://en.wikipedia.org/wiki/Test_functions_for_optimization)
provided by `torchopt` including `"ackley"`, `"beale"`, `"booth"`,
`"bukin_n6"`, `"easom"`, `"goldstein_price"`, `"himmelblau"`,
`"levi_n13"`, `"matyas"`, `"rastrigin"`, `"rosenbrock"`, `"sphere"`.
Optimization functions are useful to evaluate characteristics of
optimization algorithms, such as convergence rate, precision,
robustness, and performance. These functions give an idea about the
different situations that optimization algorithms can face.

In what follows, we perform tests using `"beale"` test function. To
visualize an animated GIF, we set `plot_each_step=TRUE` and capture each
step frame using [gifski](https://CRAN.R-project.org/package=gifski)
package.

### `optim_adamw()`:

``` r
# test optim adamw
set.seed(12345)
torchopt::test_optim(
    optim = torchopt::optim_adamw,
    test_fn = "beale",
    opt_hparams = list(lr = 0.1),
    steps = 500,
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_adamw-.gif" width="50%" height="50%" />

### `optim_adabelief()`:

``` r
set.seed(42)
test_optim(
    optim = optim_adabelief,
    opt_hparams = list(lr = 0.5),
    steps = 400,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_adabelief-.gif" width="50%" height="50%" />

### `optim_adabound()`:

``` r
# set manual seed
set.seed(22)
test_optim(
    optim = optim_adabound,
    opt_hparams = list(lr = 0.5),
    steps = 400,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_adabound-.gif" width="50%" height="50%" />

### `optim_adahessian()`:

``` r
# set manual seed
set.seed(290356)
test_optim(
    optim = optim_adahessian,
    opt_hparams = list(lr = 0.2),
    steps = 500,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_adahessian-.gif" width="50%" height="50%" />

### `optim_madgrad()`:

``` r
set.seed(256)
test_optim(
    optim = optim_madgrad,
    opt_hparams = list(lr = 0.05),
    steps = 400,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_madgrad-.gif" width="50%" height="50%" />

### `optim_nadam()`:

``` r
set.seed(2903)
test_optim(
    optim = optim_nadam,
    opt_hparams = list(lr = 0.5, weight_decay = 0),
    steps = 500,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_nadam-.gif" width="50%" height="50%" />

### `optim_qhadam()`:

``` r
set.seed(1024)
test_optim(
    optim = optim_qhadam,
    opt_hparams = list(lr = 0.1),
    steps = 500,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_qhadam-.gif" width="50%" height="50%" />

### `optim_radam()`:

``` r
set.seed(1024)
test_optim(
    optim = optim_radam,
    opt_hparams = list(lr = 1.0),
    steps = 500,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_radam-.gif" width="50%" height="50%" />

### `optim_swats()`:

``` r
set.seed(234)
test_optim(
    optim = optim_swats,
    opt_hparams = list(lr = 0.5),
    steps = 500,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_swats-.gif" width="50%" height="50%" />

### `optim_yogi()`:

``` r
# set manual seed
set.seed(66)
test_optim(
    optim = optim_yogi,
    opt_hparams = list(lr = 0.1),
    steps = 500,
    test_fn = "beale",
    plot_each_step = TRUE
)
```

<img src="man/figures/README-test_yogi-.gif" width="50%" height="50%" />

## Acknowledgements

We are thankful to Collin Donahue-Oponski <https://github.com/colllin>,
Amir Gholami <https://github.com/amirgholami>, Liangchen Luo
<https://github.com/Luolc>, Liyuan Liu
<https://github.com/LiyuanLucasLiu>, Nikolay Novik
<https://github.com/jettify>, Patrik Purgai
<https://github.com/Mrpatekful> Juntang Zhuang
<https://github.com/juntang-zhuang> and the PyTorch team
<https://github.com/pytorch/pytorch> for providing pytorch code for the
optimizers implemented in this package. We also thank Daniel Falbel
<https://github.com/dfalbel> for providing support for the R version of
PyTorch.

## Code of Conduct

The torchopt project is released with a [Contributor Code of
Conduct](https://contributor-covenant.org/version/2/0/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.

## References

-   ADABELIEF: Juntang Zhuang, Tommy Tang, Yifan Ding, Sekhar Tatikonda,
    Nicha Dvornek, Xenophon Papademetris, James S. Duncan. “Adabelief
    Optimizer: Adapting Stepsizes by the Belief in Observed Gradients”,
    34th Conference on Neural Information Processing Systems (NeurIPS
    2020), <https://arxiv.org/abs/2010.07468>.

-   ADABOUND: Liangchen Luo, Yuanhao Xiong, Yan Liu, Xu Sun, “Adaptive
    Gradient Methods with Dynamic Bound of Learning Rate”, International
    Conference on Learning Representations (ICLR), 2019.
    <https://doi.org/10.48550/arXiv.1902.09843>.

-   ADAHESSIAN: Zhewei Yao, Amir Gholami, Sheng Shen, Mustafa Mustafa,
    Kurt Keutzer, Michael W. Mahoney. “Adahessian: An Adaptive Second
    Order Optimizer for Machine Learning”, AAAI Conference on Artificial
    Intelligence, 35(12), 10665-10673, 2021.
    <https://arxiv.org/abs/2006.00719>.

-   ADAMW: Ilya Loshchilov, Frank Hutter, “Decoupled Weight Decay
    Regularization”, International Conference on Learning
    Representations (ICLR) 2019.
    <https://doi.org/10.48550/arXiv.1711.05101>.

-   MADGRAD: Aaron Defazio, Samy Jelassi, “Adaptivity without
    Compromise: A Momentumized, Adaptive, Dual Averaged Gradient Method
    for Stochastic Optimization”, arXiv preprint arXiv:2101.11075, 2021.
    <https://doi.org/10.48550/arXiv.2101.11075>

-   NADAM: Timothy Dazat, “Incorporating Nesterov Momentum into Adam”,
    International Conference on Learning Representations (ICLR), 2019.
    <https://openreview.net/pdf/OM0jvwB8jIp57ZJjtNEZ.pdf>

-   QHADAM: Jerry Ma, Denis Yarats, “Quasi-hyperbolic momentum and Adam
    for deep learning”. <https://arxiv.org/abs/1810.06801>

-   RADAM: Liyuan Liu, Haoming Jiang, Pengcheng He, Weizhu Chen,
    Xiaodong Liu, Jianfeng Gao, Jiawei Han, “On the Variance of the
    Adaptive Learning Rate and Beyond”, International Conference on
    Learning Representations (ICLR) 2020.
    <https://arxiv.org/abs/1908.03265>.

-   SWATS: Nitish Keskar, Richard Socher, “Improving Generalization
    Performance by Switching from Adam to SGD”. International Conference
    on Learning Representations (ICLR), 2018.
    <https://arxiv.org/abs/1712.07628>.

-   YOGI: Manzil Zaheer, Sashank Reddi, Devendra Sachan, Satyen Kale,
    Sanjiv Kumar, “Adaptive Methods for Nonconvex Optimization”,
    Advances in Neural Information Processing Systems 31 (NeurIPS 2018).
    <https://papers.nips.cc/paper/8186-adaptive-methods-for-nonconvex-optimization>


NEWS.md
# R Implementation of Advanced Optimizers for Torch

### What's new in torchopt version 0.1.4

* fixed errors in DESCRIPTION

### What's new in torchopt version 0.1.2

* adahessian optimizer
* nadam optimizer
* radam optimizer 
* qhadam optimizer
* swats optimizer

### What's new in torchopt version 0.1.1

* adabelief optimizer
* adabound optimizer
* adamw optimizer
* madgrad optimizer
* yogi optimizer



R/utils-testopt.R
ackley <- function(x,y) {
    -20 * exp(-0.2*sqrt(0.5*(x^2 + y^2))) - exp(0.5*(cos(2*pi*x) + cos(2*pi*y))) + exp(1) + 20
}
domain_ackley <- function(){
    x0 <- runif(1,-5, 5)
    y0 <- runif(1,-5, 5)
    return(c(x0 = x0, y0 = y0, xmax = 5, xmin = -5, ymax = 5, ymin = -5))
}

beale <- function(x, y) {
    log((1.5 - x + x * y)^2 + (2.25 - x - x * y^2)^2 + (2.625 - x + x * y^3)^2)
}
domain_beale <- function(){
    x0 <- runif(1,-4.5, 4.5)
    y0 <- runif(1,-4.5, 4.5)
    return(c(x0 = x0, y0 = y0, xmax = 4.5, xmin = -4.5, ymax = 4.5, ymin = -4.5))
}

booth <- function(x, y) {
    log((x + 2 * y - 7)^2 + (2 * x + y - 5)^2)
}
domain_booth <- function(){
    x0 <- runif(1,-10, 10)
    y0 <- runif(1,-10, 10)
    return(c(x0 = x0, y0 = y0, xmax = 10, xmin = -10, ymax = 10, ymin = -10))
}

bukin_n6 <- function(x, y) {
    100 * sqrt(abs(y - 0.01 * x^2)) + 0.01 * abs(x + 10)
}
domain_bukin_n6 <- function(){
    x0 <- runif(1,-15, -5)
    y0 <- runif(1,-3, 3)
    return(c(x0 = x0, y0 = y0, xmax = -5, xmin = -15, ymax = -3, ymin = 3))
}

easom <- function(x, y) {
    -cos(x) * cos(y) * exp(-(x - pi)^2 - (y - pi)^2)
}
domain_easom <- function(){
    x0 <- runif(1,-1, 7)
    y0 <- runif(1,-1, 7)
    return(c(x0 = x0, y0 = y0, xmax = 7, xmin = -1, ymax = 7, ymin = -1))
}
goldstein_price <- function(x, y) {
    log((1 + (x + y + 1)^2 *
             (19 - 14 * x + 3 * x^2 - 14 * y + 6 * x * y + 3 * y^2)) *
            (30 + (2 * x - 3 * y)^2 * (18 - 32 * x + 12 * x^2 + 48 *
                                           y - 36 * x * y + 27 * y^2)))
}
domain_goldstein_price <- function(){
    x0 <- runif(1,-2, 2)
    y0 <- runif(1,-3, 1)
    return(c(x0 = x0, y0 = y0, xmax = 2, xmin = -2, ymax = -3, ymin = 1))
}
himmelblau <- function(x, y) {
    log((x^2 + y - 11)^2 + (x + y^2 - 7)^2)
}
domain_himmelblau <- function(){
    x0 <- runif(1,-5, 5)
    y0 <- runif(1,-5, 5)
    return(c(x0 = x0, y0 = y0, xmax = 5, xmin = -5, ymax = 5, ymin = -5))
}
levi_n13 <- function(x, y) {
    sin(3 * pi * x)^2 + (x - 1)^2 * (1 + sin(3 * pi * y)^2) +
        (y - 1)^2 * (1 + sin(2 * pi * y)^2)
}
domain_levi_n13 <- function(){
    x0 <- runif(1,-5, 7)
    y0 <- runif(1,-5, 7)
    return(c(x0 = x0, y0 = y0, xmax = 7, xmin = -5, ymax = 7, ymin = -5))
}
matyas <- function(x, y) {
    log(0.26 * (x^2 + y^2) - 0.48 * x * y)
}
domain_matyas <- function(){
    x0 <- runif(1,-10, 10)
    y0 <- runif(1,-10, 10)
    return(c(x0 = x0, y0 = y0, xmax = 10, xmin = -10, ymax = 10, ymin = -10))
}
rastrigin <- function(x, y) {
    20 + (x^2 - 10 * cos(2 * pi * x)) + (y^2 - 10 * cos(2 * pi * y))
}
domain_rastrigin <- function(){
    x0 <- runif(1,-5.12, 5.12)
    y0 <- runif(1,-5.12, 5.12)
    return(c(x0 = x0, y0 = y0, xmax = 5.12, xmin = -5.12, ymax = 5.12, ymin = -5.12))
}
rosenbrock <- function(x, y) {
    log(100 * (y - x^2)^2 + (1 - x)^2)
}
domain_rosenbrock <- function(){
    x0 <- -2
    y0 <- 2
    return(c(x0 = x0, y0 = y0, xmax = 2, xmin = -2, ymax = 3, ymin = -1))
}
sphere <- function(x, y) {
    x^2 + y^2
}
domain_sphere <- function(){
    x0 <- runif(1,-2, 2)
    y0 <- runif(1,-2, 2)
    return(c(x0 = x0, y0 = y0, xmax = 2, xmin = -2, ymax = 2, ymin = -2))
}
#' @title Test optimization function
#'
#' @name test_optim
#'
#' @author Rolf Simoes, \email{rolf.simoes@@inpe.br}
#'
#' @description
#' `test_optim()` function is useful to visualize how optimizers solve the
#' minimization problem by showing the convergence path using a test function.
#' User can choose any test optimization
#' [functions](https://en.wikipedia.org/wiki/Test_functions_for_optimization)
#' provided by `torchopt`:
#'
#' `"beale"`, `"booth"`, `"bukin_n6"`, `"easom"`, `"goldstein_price"`,
#' `"himmelblau"`, `"levi_n13"`, `"matyas"`, `"rastrigin"`,
#' `"rosenbrock"`, and `"sphere"`.
#'
#' Besides these functions, users can pass any function that receives two
#' numerical values and returns a scalar.
#'
#' Optimization functions are useful to evaluate characteristics of optimization
#' algorithms, such as convergence rate, precision, robustness, and performance.
#' These functions give an idea about the different situations that optimization
#' algorithms can face.
#'
#' Function `test_function()` plot the 2D-space of a test optimization function.
#'
#' @param optim          Torch optimizer function.
#' @param ...            Additional parameters (passed to `image` function).
#' @param opt_hparams    A list with optimizer initialization parameters (default: `list()`).
#' If missing, for each optimizer its individual defaults will be used.
#' @param test_fn        A test function (default `"beale"`). You can also pass
#'   a list with 2 elements. The first should be a function that will be optimized
#'   and the second is a function that returns a named vector with `x0`, `y0`
#'   (the starting points) and `xmax`, `xmin`, `ymax` and `ymin` (the domain).
#'   An example: `c(x0 = x0, y0 = y0, xmax = 5, xmin = -5, ymax = 5, ymin = -5)`
#' @param steps          Number of steps to run (default `200`).
#' @param pt_start_color Starting point color (default `"#5050FF7F"`)
#' @param pt_end_color   Ending point color (default `"#FF5050FF"`)
#' @param ln_color       Line path color (default `"#FF0000FF"`)
#' @param ln_weight      Line path weight (default `2`)
#' @param bg_xy_breaks   Background X and Y resolution (default `100`)
#' @param bg_z_breaks    Background Z resolution (default `32`)
#' @param bg_palette     Background palette (default `"viridis"`)
#' @param ct_levels      Contour levels (default `10`)
#' @param ct_labels      Should show contour labels? (default `FALSE`)
#' @param ct_color       Contour color (default `"#FFFFFF7F"`)
#' @param plot_each_step Should output each step? (default `FALSE`)
#'
#' @return No return value, called for producing animated gifs
#'
#' @export
test_optim <- function(optim, ...,
                       opt_hparams = list(),
                       test_fn = "beale",
                       steps = 200,
                       pt_start_color = "#5050FF7F",
                       pt_end_color = "#FF5050FF",
                       ln_color = "#FF0000FF",
                       ln_weight = 2,
                       bg_xy_breaks = 100,
                       bg_z_breaks = 32,
                       bg_palette = "viridis",
                       ct_levels = 10,
                       ct_labels = FALSE,
                       ct_color = "#FFFFFF7F",
                       plot_each_step = FALSE) {

    # pre-conditions
    inherits_from <- if (utils::packageVersion("torch") > '0.7.2') "torch_optimizer_generator" else "function"
    if (!inherits(optim, inherits_from)) {

        stop("invalid 'optim' param.", call. = FALSE)
    }
    if (is.character(test_fn)) {
        if (!exists(test_fn,
                    envir = asNamespace("torchopt"),
                    inherits = FALSE)) {
            stop("invalid 'test_fn' param.", call. = FALSE)
        }
        # get starting points
        domain_fn <- get(paste0("domain_",test_fn),
                         envir = asNamespace("torchopt"),
                         inherits = FALSE)
        # get gradient function
        test_fn <- get(test_fn,
                       envir = asNamespace("torchopt"),
                       inherits = FALSE)
    } else if (is.list(test_fn)) {
        domain_fn <- test_fn[[2]]
        test_fn <- test_fn[[1]]
    }

    if (!is.function(test_fn)) {
        stop("invalid 'test_fn' param.", call. = FALSE)
    }
    if (!is.function(domain_fn)) {
        stop("missing domain param for function.", call. = FALSE)
    }
    # starting point
    dom <- domain_fn()
    x0 <- dom[["x0"]]
    y0 <- dom[["y0"]]
    # create tensor
    x <- torch::torch_tensor(x0, requires_grad = TRUE)
    y <- torch::torch_tensor(y0, requires_grad = TRUE)

    # instantiate optimizer
    optim <- do.call(optim, c(list(params = list(x, y)), opt_hparams))
    grad_keep <-  FALSE
    if (inherits(optim, "optim_adahessian")) {
        grad_keep <- TRUE
        # retain_graph is not exposed before torch 0.7.2
        if (!utils::packageVersion("torch") > '0.7.2') {
            stop("adahessian needs torch version > 0.7.2, got ",
                 utils::packageVersion("torch"))
        }
    }
    # run optimizer
    x_steps <- numeric(steps)
    y_steps <- numeric(steps)
    for (i in seq_len(steps)) {
        x_steps[i] <- as.numeric(x)
        y_steps[i] <- as.numeric(y)
        optim$zero_grad()
        z <- test_fn(x, y)
        # retain_graph is not exposed before torch 0.7.2
        if (utils::packageVersion("torch") > '0.7.2') {
            z$backward(create_graph = grad_keep, retain_graph = grad_keep)
        } else {
            z$backward(create_graph = grad_keep)
        }
        optim$step()
    }

    # prepare plot
    # get xy limits

    xmax <- dom[["xmax"]]
    xmin <- dom[["xmin"]]
    ymax <- dom[["ymax"]]
    ymin <- dom[["ymin"]]

    # prepare data for gradient plot
    x <- seq(xmin, xmax, length.out = bg_xy_breaks)
    y <- seq(xmin, xmax, length.out = bg_xy_breaks)
    z <- outer(X = x, Y = y, FUN = function(x, y) as.numeric(test_fn(x, y)))

    plot_from_step <- steps
    if (plot_each_step) {
        plot_from_step <- 1
    }

    for (step in seq(plot_from_step, steps, 1)) {

        # plot background
        image(
            x = x,
            y = y,
            z = z,
            col = hcl.colors(
                n = bg_z_breaks,
                palette = bg_palette
            ),
            ...
        )

        # plot contour
        if (ct_levels > 0) {
            contour(
                x = x,
                y = y,
                z = z,
                nlevels = ct_levels,
                drawlabels = ct_labels,
                col = ct_color,
                add = TRUE
            )
        }

        # plot starting point
        points(
            x_steps[1],
            y_steps[1],
            pch = 21,
            bg = pt_start_color
        )

        # plot path line
        lines(
            x_steps[seq_len(step)],
            y_steps[seq_len(step)],
            lwd = ln_weight,
            col = ln_color
        )

        # plot end point
        points(
            x_steps[step],
            y_steps[step],
            pch = 21,
            bg = pt_end_color
        )
    }
}


